File: dorm-service/model/dorm.go
package model

type Dorm struct {
	ID          string   `json:"id"` // može biti UUID ili ObjectID
	Name        string   `json:"name"`
	Address     string   `json:"address"`
	City        string   `json:"city"`
	Price       float64  `json:"price"`
	Capacity    int      `json:"capacity"`
	Occupied    int      `json:"occupied"`  // trenutno zauzetih mesta
	Type        string   `json:"type"`      // npr. "muški", "ženski", "mešoviti"
	Amenities   []string `json:"amenities"` // npr. "WiFi", "kantina", "biblioteka"
	Description string   `json:"description,omitempty"`

	// Korisnički aspekti
	Ratings     []Rating  `json:"ratings,omitempty"`
	Comments    []Comment `json:"comments,omitempty"`
	Preferences []string  `json:"preferences,omitempty"`
	Users       []string  `json:"users,omitempty" bson:"users,omitempty"`

	FreeSpots int `bson:"-" json:"free_spots"`
}

type Rating struct {
	UserID string  `json:"user_id" bson:"user_id"`
	Score  float64 `json:"score" bson:"score"`
}

type Comment struct {
	UserID  string `json:"user_id"`
	Message string `json:"message"`
	Date    string `json:"date"` // ISO 8601 format (npr. "2025-09-12T14:30:00Z")
}

File: student-dorms-frontend/src/app/services/dorm.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { AuthService } from './auth';

export interface Dorm {
  id: string;
  name: string;
  address: string;
  city: string;
  capacity: number;
  price: number;
  occupied: number;
  type: string;
  amenities: string[];
  description?: string;
  ratings?: { user_id: string; score: number }[];
  comments?: { user_id: string; message: string; date: string }[];
  preferences?: string[];
}

@Injectable({
  providedIn: 'root'
})
export class DormService {
  private apiUrl = 'http://localhost:8000/dorms'; 

  constructor(private http: HttpClient, private authService: AuthService) {
    
  }

  createDorm(dorm: Dorm): Observable<any> {
    const token = this.authService.getToken(); // ili localStorage.getItem('token')
    const headers = token ? new HttpHeaders({ Authorization: `Bearer ${token}` }) : undefined;

    return this.http.post(`${this.apiUrl}/dorms`, dorm, { headers });
  }

  getDorms(): Observable<Dorm[]> {
    return this.http.get<Dorm[]>(this.apiUrl);
  }

  getDormById(id: string): Observable<Dorm> {
    return this.http.get<Dorm>(`${this.apiUrl}/dorms/${id}`);
  }

  getAllDorms(): Observable<Dorm[]> {
    return this.http.get<Dorm[]>(`${this.apiUrl}/dorms`);
  }

  addRating(dormId: string, score: number): Observable<any> {
    const token = localStorage.getItem('token'); 

    const headers = new HttpHeaders({
      Authorization: `Bearer ${token}`
    });

    return this.http.post(
      `${this.apiUrl}/dorms/rating`,
      { dorm_id: dormId, score: score },
      { headers }
    );
  }

}
